Tracing filter

JonathanWoollett-Light opened the pull request:
## Changes

Add the ` FilterFn` layer to support filtering by crate, module and span.
With `tracing::instrument` spans wrap functions, this would
support filtering to specific functions.

## Reason

...

## Status

Passes all tests except `test_api_requests_logs`. If the filter function is simply changed to always return `true` it passes this test, implying the issue relates to the filter being incorrect. I was thinking the filter level doesn't get updated from the default (`Warn`) to `Info` in time to print the log message, but adding a timeout between the API request to update the log level and the following API request where it expects to see the log doesn't appear to make any difference.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

JonathanWoollett-Light commented on src/firecracker/src/main.rs:
```diff
@@ -292,19 +290,20 @@ fn main_exitable() -> FcExitCode {
+            show_log_origin: Some(arguments.flag_present("show-log-origin")),
         };
-    }
+        match config.init() {
+            Ok(h) => h,
+            Err(err) => {
+                return generic_error_exit(&format!("Could not initialize logger: {}", err));
+            }
+        }
```
I might be missing what you are suggesting. I'm thinking you can't return from within a closure so I'm not sure what your suggested change is here.

ShadowCurse replied:
Oh yeah, I was mistaken here.

JonathanWoollett-Light commented on src/vmm/src/rpc_interface.rs:
```diff
@@ -1257,320 +1254,13 @@ mod tests {
         Ok(Arc::new(Mutex::new(MockVmm::default())))
     }
 
-    fn default_preboot<'a>(
-        vm_resources: &'a mut VmResources,
-        event_manager: &'a mut EventManager,
```
These tests do cannot work without mocks and in the new implementation almost entirely just test that mock code. A significant proportion of the coverage differnece is a result of this.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/mod.rs:
```diff
@@ -21,8 +36,6 @@ pub mod drive;
 /// Wrapper over the microVM general information attached to the microVM.
 pub mod instance_info;
-/// Wrapper for configuring the logger.
-pub mod logger;
```
Fixed.

bchalios commented on src/vmm/src/vmm_config/mod.rs:
```diff
@@ -177,6 +183,258 @@ fn open_file_nonblock(path: &Path) -> Result<File, std::io::Error> {
+/// Strongly typed structure used to describe the logger.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct LoggerConfig {
```
It would also help the review if all of these types remained in the `logger` module so we could see the diff.

JonathanWoollett-Light replied:
I've removed the changes that moved it out of the `logger.rs` file.

bchalios commented on src/api_server/Cargo.toml:
```diff
@@ -21,7 +21,7 @@ mmds = { path = "../mmds" }
 utils = { path = "../utils" }
 vmm = { path = "../vmm" }
-tracing = { version = "0.1.37", features = ["attributes"] }
+tracing = { git = "https://github.com/tokio-rs/tracing", rev = "27f688efb72316a26f3ec1f952c82626692c08ff", features = ["attributes"] }
```
Why?!? I really think we should not be doing this.

JonathanWoollett-Light replied:
The latest version of tracing published on crates.io is 10 months ago. The main branch is up to date and includes signficant changes, I beleive this is more secure and future proof.

JonathanWoollett-Light replied:
> Why more secure? I would actually argue that this is less secure. Less projects are using this commit, comparing to the published version, downstream so it has gone through less testing in the field.

It is using more up to date dependencies with more up to date fixes, the majority of crates do not report when they have errors, they only fix them in version updates. In this case a more up to date version may be more secure.

In the case a security issue is found in any dependencies in the tree, we would need to do the large update from the last version onto main to resolve the issue.

> The main branch is a branch which means it can be re-written, which means this commit can disappear. I know it's very unlikely but it can happen.

If we consider this a significant risk we can fork `tracing` and depend on this fork.

> Is there a change included there that we really need and cannot implement our functionality otherwise?

I did not see the ability to re-configure the logger which underlies much of the simplification of our implementation in the older version publish to crates.io

bchalios replied:
> In the case a security issue is found in any dependencies in the tree, we would need to do the large update from the last version onto main to resolve the issue.

Do you mean that if they find a security issue they will not cut a new release? If that's the case we should not be using this crate.

bchalios replied:
> I was not able see the ability to re-configure the logger which underlies much of the simplification of our implementation in the older version published to crates.io.

That is problematic. Again, I think we need input from more people.

JonathanWoollett-Light replied:
> Do you mean that if they find a security issue they will not cut a new release? If that's the case we should not be using this crate.

No they may cut a release, but they may not cut a SVE, in this case `cargo audit` wont catch it and notify us and tracing won't publish a new version. But the latest version of tracing will always update to the latest dependencies. Thus the version on main will be safe will the older version will not be.

bchalios replied:
That is the case with every external crate we consume. Should we point to their git history and update the hash every time the push something new?

My point is that pointing to a git has does not solve any security problem.

bchalios replied:
Also, they may not cut as an SVE, but if they publish a new version we will get a dependabot PR for it.

bchalios commented on src/firecracker/src/main.rs:
```diff
@@ -119,7 +119,7 @@ fn main_exitable() -> FcExitCode {
             Argument::new("id")
                 .takes_value(true)
-                .default_value(logger::DEFAULT_INSTANCE_ID)
+                .default_value("anonymous-instance")
```
Why not use a `const` and use a string literal?

JonathanWoollett-Light replied:
It's only used here so making a constant only adds additional lines.

bchalios commented on src/firecracker/src/main.rs:
```diff
@@ -201,6 +198,11 @@ fn main_exitable() -> FcExitCode {
+            Argument::new("new-format")
+                .takes_value(false)
+                .help("Whether to use the new logging output format."),
+        )
```
You seem to add this in the first commit and remove it later. Why?

JonathanWoollett-Light replied:
The 1st commit supported the old logger format and this format could be changed with this argument. The later commit removes the old format as in discussion we saw no reason to maintain the old format.

bchalios commented on tests/integration_tests/functional/test_api_server.py:
```diff
@@ -24,5 +24,5 @@ def test_api_socket_in_use(test_microvm_with_api):
     sock.bind(microvm.jailer.api_socket_path())
     microvm.spawn()
-    msg = "Failed to open the API socket at: /run/firecracker.socket. Check that it is not already used."
+    msg = "Firecracker panicked at 'The channel's sending half was disconnected."
```
Is that due to the intermittent CI error we had?

bchalios replied:
`s/had/have`

JonathanWoollett-Light replied:
I do not think its the same issue, but it is similar.

JonathanWoollett-Light replied:
I've removed this change.

bchalios commented on tests/integration_tests/build/test_coverage.py:
```diff
@@ -25,9 +25,9 @@ def is_on_skylake():
 if utils.is_io_uring_supported():
-    COVERAGE_DICT = {"Intel": 82.44, "AMD": 81.68, "ARM": 80.95}
+    COVERAGE_DICT = {"Intel": 80.41, "AMD": 79.59, "ARM": 78.77}
 else:
-    COVERAGE_DICT = {"Intel": 79.76, "AMD": 78.90, "ARM": 78.01}
+    COVERAGE_DICT = {"Intel": 77.72, "AMD": 76.79, "ARM": 75.82}
```
That's a significant drop. Any idea why it's happening?

JonathanWoollett-Light commented on src/logger/src/lib.rs:
```diff
@@ -6,16 +6,12 @@
 mod metrics;
 
-use std::sync::LockResult;
+use std::sync::OnceLock;
```
I think these where used to acheive similar results.

`OnceLock` is the best practice way to do it.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/mod.rs:
```diff
@@ -177,6 +183,258 @@ fn open_file_nonblock(path: &Path) -> Result<File, std::io::Error> {
+    /// Named pipe or file used as output for logs.
+    pub log_path: std::path::PathBuf,
+    /// The level of the Logger.
+    pub level: Option<Level>,
```
All API calls that previously work will still work, we are not breaking the API. This is more of an implementation detail. Rather than having serde insert a default value if one isn't found when deserializing, we allow it to be `None` then we insert the default value with `.unwrap_or(MY_DEFAULT)`.

JonathanWoollett-Light commented on src/vmm/src/devices/virtio/vsock/test_utils.rs:
```diff
@@ -209,7 +210,7 @@ pub fn read_packet_data(pkt: &VsockPacket, mem: &GuestMemoryMmap, how_much: usiz
 impl<B> Vsock<B>
 where
-    B: VsockBackend,
+    B: VsockBackend + Debug,
```
I think this is left-over and should've been merged with https://github.com/firecracker-microvm/firecracker/pull/3807. I can drop it from this PR and add it back when needed (if/when adding instrumentation that needs it).

JonathanWoollett-Light replied:
I've dropped it from this PR.

JonathanWoollett-Light commented on src/vmm/src/resources.rs:
```diff
@@ -856,18 +856,15 @@ mod tests {
+                None,
+            ),
+            Err(Error::Logger(LoggerConfigError::File(_))),
+        ));
```
Previously the logger could only be configured once, after which the API would return errors if you tried to re-intialize it. Now the logger can be configured, we don't need to disable it here.

JonathanWoollett-Light replied:
This comment was wrong.
```rust
LOGGER.set_max_level(LevelFilter::Off);
```
doesn't disable the logger (it doesn't close file descriptors to its write targets or drop itself) it only changes the filter log level.

I think previously it would first check the log level before attempting to take action, thus would hit the logger file error (otherwise it would hit the re-intialization error).

SInce the logger now allows re-configuration this error case doesn't happen so it hits the file error without needing this adjustment.

JonathanWoollett-Light commented on src/api_server/src/parsed_request.rs:
```diff
@@ -3,10 +3,10 @@
 
 use std::fmt::Debug;
 
-use logger::{error, info, log_enabled, Level};
```
`logging::error` used to re-export `log::error`  (e.g. `pub use log::error;`). We could also re-export `tracing::error` but I don't like this, I find re-exports add complexity and reduce readability.

I understand your point if we wanted to change away from `tracing` but I don't think it's valuable for this PR to be designed considering it may be reverted.

I think using `tracing::error` (etc.) is the simplest and best approach (I think if we avoid this with wrapping here, the same point could be argued in many cases and not be useful).

bchalios replied:
I don't have a super strong opinion here, but I do not understand how re-exporting adds complexity or reduces readability. Re-exporting just renames something. Also, it removes the need to "polute" Cargo.toml with both `logger` and `tracing`.

I think it's worth getting more opinions here

roypat replied:
Mh, I can see a few (weak) arguments for re-exporting
- It would prevent us from having to add `tracing` as a dependency to a bunch of `Cargo.toml` files (however we're in the process of merging all crates, so this will resolve itself either way)
- The diff for this commit wouldn't include changes of the type `log::error!` -> `tracing::error` (however we dont tend to fully qualify our logging calls, instead mostly opting for importing the macros at the top of the file).

Overall I think reexporting would be nicer (diff would be a bit smaller, and we avoid having to specify the dependency a ton of times), but I also don't have a strong opinion.

JonathanWoollett-Light replied:
I've updated it to maintain the same re-exporting pattern.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/mod.rs:
```diff
@@ -177,6 +183,258 @@ fn open_file_nonblock(path: &Path) -> Result<File, std::io::Error> {
+            .map_err(LoggerConfigError::File)?;
+
+        // Write the initialization message.
+        file.write_all(LoggerConfig::INIT_MESSAGE.as_bytes())
```
I've fixed it in this commit (we really should automate testing this in CI).

JonathanWoollett-Light commented:
@bchalios to address points covered in your comments but not comments on code:

> Finally, this PR seems to just rework our logger in order to use tracing for logging. Will actual tracing capabilities be in subsequent PR?

The additional capabilities are a little more controversial it seems so will require some more justifcation. I think @pb8o is not entirely convinced of their use.

> Could you please rework the commit message? This is not a fix, it's probably a new feature. Also, could you please explain in the body why we're making this change and what is tracing_subscriber.

I've changed the commit title to ``feat: Replace `log` with `tracing` `` and updated the description.

> Cool, but it doesn't seem to fit in this PR.

I can drop it from this PR but I'm not sure the effort to create a seperate PR for this is worth it.

> What's the purpose of this? Explanation in the commit message is missing. Also, I assume this needs documentation(?)

I've added more details to the commit body. This filter is uses an enviroment when compiling FIrecracker, it is not configurable by a user of the binary.

bchalios commented on src/firecracker/src/main.rs:
```diff
@@ -122,7 +119,7 @@ fn main_exitable() -> FcExitCode {
             Argument::new("id")
                 .takes_value(true)
-                .default_value(DEFAULT_INSTANCE_ID)
+                .default_value(logger::DEFAULT_INSTANCE_ID)
```
Just make this a string literal in the first commit. Right now you do

```
-                .default_value(DEFAULT_INSTANCE_ID)
+                .default_value(logger::DEFAULT_INSTANCE_ID)
```

and then later:

```
-                .default_value(logger::DEFAULT_INSTANCE_ID)
+                .default_value("anonymous-instance")
```

bchalios reviewed (commented):
Is the `EnvFilter` functionality specific to `tracing::instrument` as the commit message seems to imply? If that is the case, this should not be included in this PR.

In any case, this change needs documentation on how to use it during compilation time. Also, probably hooks in `devtool` to do so

bchalios commented:
Hmmm, Github doesn't like me reviewing per commit :( It discarded half of my comments.

General comments:

* I think you can squash the first 3 commits.
* There are things you change in the 1st commit and then you cancel the change in the 4th commit. It'd be nice if the 4th commit only did the changes needed for making the logger re-configurable. All the generic refactoring of the logger should go to the 1st commit. I've tried to comment those wherever I spotted them.

I have more questions regarding the logger re-configurability:
1. What is the use case for it?
2. IIUC the API for re-configuring the logger is sending a `PUT` request in the `/logger` endpoint. I think the proper way of doing this is making the `/logger` endpoint handling `PATCH` requests as well.
3. If we end-up doing this, this PR needs to add documentation and change the CHANGELOG.

bchalios commented on src/logger/src/lib.rs:
```diff
@@ -30,6 +26,11 @@ pub type FcLineWriter = std::io::LineWriter<std::fs::File>;
+/// The default instance ID.
+pub const DEFAULT_INSTANCE_ID: &str = "anonymous-instance";
+/// The instance ID to use when initializing the logger.
+pub static INSTANCE_ID: OnceLock<String> = OnceLock::new();
```
Another, temporary change that I spotted:

You add the `INSTANCE_ID: OnceLock<String>` type in the first commit and then remove it all together.

JonathanWoollett-Light replied:
I've squashed d4b7864be15ce89a82a5e700e9afdf8b1307d275 and b1ccbcc21e918e4fcad4e5bd22f0ab295b60460a. I think this should resolve these issues.

bchalios replied:
I liked them separate :( I just wanted to have each commit do exactly what it claims to do.

JonathanWoollett-Light commented:
> Is the `EnvFilter` functionality specific to `tracing::instrument` as the commit message seems to imply? If that is the case, this should not be included in this PR.
> 
> In any case, this change needs documentation on how to use it during compilation time. Also, probably hooks in `devtool` to do so

`tracing::instrument` is just a macro that generates spans/logs, `EnvFilter` can be used for all spans/logs.

Documentation on how to use it can be seen https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html

JonathanWoollett-Light commented:
> I have more questions regarding the logger re-configurability:
> 1. What is the use case for it?
> 2. IIUC the API for re-configuring the logger is sending a `PUT` request in the `/logger` endpoint. I think the proper way of doing this is making the `/logger` endpoint handling `PATCH` requests as well.
> 3. If we end-up doing this, this PR needs to add documentation and change the CHANGELOG.

The basic use case is to support better design where it always initializes the logger when the firecracker process starts targeting logs to stdout. Previously the logger wouldn't be initialized but when we logged error or warn level logs it would handle this differently and log to stderr. With this new approach we no longer need this special handling of specific logs based on level (to a specific write target).

Other application may be that you care about logs for a specific API call so configure the logger differently for this API call to output more or less verbose logs, or to a different target.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/logger.rs:
```diff
@@ -1,257 +1,298 @@
-            LoggerLevel::from_string("DEBUG".to_string()).unwrap(),
-            LoggerLevel::Debug
-        );
+static SHOW_LEVEL: AtomicBool = AtomicBool::new(false);
+static SHOW_LOG_ORIGIN: AtomicBool = AtomicBool::new(false);
```
`SHOW_LEVEL` could be updated to provide more verbose data about a specific API call a user might be interested in.

bchalios replied:
> I think this is fine, I can't think of a circumstance where this is an issue.

It might be an issue if a user is expecting particular log formats at every given point in time.

> SHOW_LEVEL could be updated to provide more verbose data about a specific API call a user might be interested in.

I'm not sure how someone can do that per API call, based on your changes. Could you give an example?

JonathanWoollett-Light replied:
```bash
# Default log level is `Warn` update to `Debug` to be more verbose.
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Debug\",
    }" \
    "http://localhost/logger"

# Set boot source
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"kernel_image_path\": \"${KERNEL}\",
        \"boot_args\": \"${KERNEL_BOOT_ARGS}\"
    }" \
    "http://localhost/boot-source"

# Reduce log verbosity back down to `Warn`
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Warn\",
    }" \
    "http://localhost/logger"

```

bchalios commented:
> The basic use case is to support better design where it always initializes the logger when the firecracker process starts targeting logs to stdout. Previously the logger wouldn't be initialized but when we logged error or warn level logs it would handle this differently and log to stderr. With this new approach we no longer need this special handling of specific logs based on level (to a specific write target).

I like that. I think it's better design. _It is_ a breaking change though. This should be documented. Particularly the part that if no logger is created users won't see `warn` & `error` messages in `stderr`.

> Other application may be that you care about logs for a specific API call so configure the logger differently for this API call to output more or less verbose logs, or to a different target.

Does this PR allow that? IIUC what this PR provide towards that direction is that now we can control this through environment variables per crate at compile time but not per API call.

> > IIUC the API for re-configuring the logger is sending a PUT request in the /logger endpoint. I think the proper way of doing this is making the /logger endpoint handling PATCH requests as well.

> This would be a breaking change to the API.

Adding a new optional endpoint [is not a breaking change](https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md#what-is-a-breaking-change)

> I also don't think any customers care if its a PUT instead of PATCH. In my opinion it would be best if all our requests where just POST and we ignored the request method as its additional complexity that offers no benefits, if we look at POSTs definition (from https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods) its general enough to apply to everything:

Maybe, but they currently our APIs are not `POST`s. And the way we use for other endpoints to update configuration at runtime is through `PATCH` requests. I think we should handle this case the same way.

> I think there are only 3 customer visible changes that can be documented:
We should also add a section on how one can use environment variables (which and how) to provide different levels of logging per crate.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/logger.rs:
```diff
@@ -1,257 +1,298 @@
+///    accepted `Warning`).
+/// 2. Setting the default to `Warn` to avoid a breaking change.
+///
+/// This alias, custom `Default` and type should be removed in the next breaking update to simplify
+/// the code and API (and `log::Level` should be used in place).
```
You're right, a reminder for firecracker 2.0

JonathanWoollett-Light commented on src/vmm/src/vmm_config/logger.rs:
```diff
@@ -1,257 +1,298 @@
 pub struct LoggerConfig {
     /// Named pipe or file used as output for logs.
-    pub log_path: PathBuf,
+    pub log_path: Option<std::path::PathBuf>,
```
No, this is not a breaking change.

JonathanWoollett-Light commented:
> I like that. I think it's better design. It is a breaking change though. This should be documented. Particularly the part that if no logger is created users won't see warn & error messages in stderr.

Ah yes you are right it is breaking. Now 4 points to document.

    1. The logger will initialize on process start writing to stdout.
    2. The logger can be re-configured by calling the API path multiple times.
    3. The log format has changed.
    4. The logger will no longer log errors and warnings to stderr.



> Does this PR allow that? IIUC what this PR provide towards that direction is that now we can control this through environment variables per crate at compile time but not per API call.

I've given an example in https://github.com/firecracker-microvm/firecracker/pull/3971#discussion_r1281767222.

JonathanWoollett-Light commented on src/api_server/src/parsed_request.rs:
```diff
@@ -229,19 +229,7 @@ fn log_received_api_request(api_description: String) {
-                )
-            }
-        }
+        ("/cpu-config", Some(payload_value)) => describe_with_body(method, path, payload_value),
```
`log_enabled!` does not work with tracing.

bchalios commented:
> You are right, I was wrong, I am thinking however this would then require we maintain 2 API endpoints, where the 2nd endpoint supersedes the 1st (and go through the effort of communicating this, then deprecating the 1st). I'm not sure its worth the effort to change the request method.

We do that already for `/drive` and `/network-interface`, I don't think it's a problem. And I don't think we need to deprecate the `PUT` method. I think the fact that the you needed to make the log-path optional in the `PUT` method, shows that we really should have two methods. This is what we do for other endpoints as well.

JonathanWoollett-Light commented on src/vmm/src/vmm_config/logger.rs:
```diff
@@ -1,257 +1,298 @@
+        let (fmt, fmt_handle) = {
+            let fmt_writer = match &self.log_path {
+                Some(path) => {
+                    // TODO Can we add `.create(true)` so the user doesn't need to pre-create the
```
I've created https://github.com/firecracker-microvm/firecracker/pull/4028 to do this.

JonathanWoollett-Light commented:
Blocked on https://github.com/firecracker-microvm/firecracker/pull/3808

JonathanWoollett-Light commented:
Moved to feature branch to store work for historical reference https://github.com/firecracker-microvm/firecracker/tree/feature/tokio-tracing.
